# 3. Feature Engineering and data preparation
* Now we have performed EDA and done preliminary cleaning of the data we can now start preparing the data ready for usage by our chosen algorithm
         
* Going to restructure data to be supervised and create some features which may be useful in aiding model to pick up the underlying relationships
   
# Imports


In [27]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
from tqdm.notebook import tqdm
from datetime import datetime

# pd.display optioons

In [28]:
pd.set_option('display.max_rows', 50)
pd.set_option('display.max_columns', 50)

# loading data

In [29]:
df_energy = pd.read_pickle("../Data/Preprocessed_Data/consumption_clustered.pkl")
df_energy

meter_id  2017-01-01  2017-01-02  \
0     0xa62b9f23553ff183f61e2bf943aab3d5983d02d7         NaN         NaN   
1     0x459c834d1f6cfb5b734b82aa9f5410fa97fb70da         NaN         NaN   
2     0x4a1ed36825360a058cec2bdd409fc2459e1ce54f         NaN         NaN   
3     0x5b76d3c0e0aefc6e0a8d1d031f96388a23263407         NaN         NaN   
4     0x943ebe39ef2be6ef807c42c5a647e27112ca5b0f         NaN         NaN   
...                                          ...         ...         ...   
3243  0x7dd7a7b8ee1bec7c44b24f738c752482f6161065       2.317       2.301   
3244  0xfdaf9f857621ec06f2cf801f42a020a322835090      14.437      16.274   
3245  0xd28f2f001e0cd4d6c121a3cb2e1427207e170e18       7.824       7.517   
3246  0x47218b46abb2fcaade487a211911406dc6e13730      23.965      28.689   
3247  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd      11.302      14.178   

      2017-01-03  2017-01-04  2017-01-05  2017-01-06  2017-01-07  2017-01-08  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243       2.352       2.516       2.229       2.354       2.397       2.397   
3244       7.031      17.018      17.603      15.005       8.987       8.490   
3245       5.398       6.788       7.360       6.898       7.321       8.042   
3246      27.664      29.229      29.548      27.909      26.923      21.277   
3247      15.499      11.853      17.431      14.506      12.812      10.472   

      2017-01-09  2017-01-10  2017-01-11  2017-01-12  2017-01-13  2017-01-14  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243       2.269      2.3040       2.339       2.212       2.456       2.229   
3244      10.136     10.3010      10.466       8.177      11.929       8.261   
3245       8.207      6.8355       5.464       7.545       6.432       6.001   
3246      23.452     24.6880      25.924      26.834      30.346      30.367   
3247      10.879     10.3525       9.826      11.969      14.236      15.300   

      2017-01-15  2017-01-16  2017-01-17  2017-01-18  2017-01-19  2017-01-20  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243       2.503       2.207       2.469       2.266       2.321       2.262   
3244      12.369      14.352      14.058      10.472      17.396       9.879   
3245       5.558       6.391       6.066       6.443       6.675       7.363   
3246      26.135      26.842      25.588      30.267      28.921      27.530   
3247      14.199       9.955      16.184      15.341      14.343      17.231   

      2017-01-21  2017-01-22  2017-01-23  2017-01-24  ...  2017-12-08  \
0            NaN         NaN         NaN         NaN  ...      3.7410   
1            NaN         NaN         NaN         NaN  ...     23.4690   
2            NaN         

In [30]:
df_addInfo = pd.read_pickle("../Data/Preprocessed_Data/addInfo_knn.pkl")
df_addInfo

meter_id  num_bedrooms  \
0     0x02b7418b821b2edeeb2baa23ba1580566b4b0e43           3.0   
1     0xaf32b2b497e9eb9c3f3de907aedc6a7914a11f41           2.0   
2     0xdd45ab28a7a40eadc04607d6da42e5ad923b1339           3.0   
3     0x2b538a5ebfc3780aef25227267063cb6410d943b           4.0   
4     0xee032ed498c9e99a4c65e0f39896fb30e513c6b7           3.0   
...                                          ...           ...   
3243  0xdf6a54b92985398af03423426e744c6d40ee8c69           4.0   
3244  0x6b6ddc9a73405caab50c8d5575b72288aaa7e844           3.0   
3245  0xd67e6b5f31d5286a2d4eec159b972e7f2197a67b           3.0   
3246  0x7dd7a7b8ee1bec7c44b24f738c752482f6161065           3.0   
3247  0x47218b46abb2fcaade487a211911406dc6e13730           3.0   

            dwelling_type  dwelling_type_detached_house  dwelling_type_flat  \
0     semi_detached_house                           0.0                 0.0   
1          terraced_house                           0.0                 0.0   
2     semi_detached_house                           0.0                 0.0   
3          detached_house                           1.0                 0.0   
4     semi_detached_house                           0.0                 0.0   
...                   ...                           ...                 ...   
3243  semi_detached_house                           0.0                 0.0   
3244       detached_house                           1.0                 0.0   
3245  semi_detached_house                           0.0                 0.0   
3246  semi_detached_house                           0.0                 0.0   
3247       detached_house                           1.0                 0.0   

      dwelling_type_semi_detached_house  dwelling_type_terraced_house  
0                                   1.0                           0.0  
1                                   0.0                           1.0  
2                                   1.0                           0.0  
3                                   0.0                           0.0  
4                                   1.0                           0.0  
...                                 ...                           ...  
3243                                1.0                           0.0  
3244                                0.0                           0.0  
3245                                1.0                           0.0  
3246                                1.0                           0.0  
3247                                0.0                           0.0  

[3248 rows x 7 columns]

In [31]:
# joining based on meter_id to assign labels to the original dataset (with all meter readings)
df_energy_addInfo = pd.merge(df_energy, df_addInfo, on="meter_id",how="inner")
df_energy_addInfo

meter_id  2017-01-01  2017-01-02  \
0     0xa62b9f23553ff183f61e2bf943aab3d5983d02d7         NaN         NaN   
1     0x459c834d1f6cfb5b734b82aa9f5410fa97fb70da         NaN         NaN   
2     0x4a1ed36825360a058cec2bdd409fc2459e1ce54f         NaN         NaN   
3     0x5b76d3c0e0aefc6e0a8d1d031f96388a23263407         NaN         NaN   
4     0x943ebe39ef2be6ef807c42c5a647e27112ca5b0f         NaN         NaN   
...                                          ...         ...         ...   
3243  0x7dd7a7b8ee1bec7c44b24f738c752482f6161065       2.317       2.301   
3244  0xfdaf9f857621ec06f2cf801f42a020a322835090      14.437      16.274   
3245  0xd28f2f001e0cd4d6c121a3cb2e1427207e170e18       7.824       7.517   
3246  0x47218b46abb2fcaade487a211911406dc6e13730      23.965      28.689   
3247  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd      11.302      14.178   

      2017-01-03  2017-01-04  2017-01-05  2017-01-06  2017-01-07  2017-01-08  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243       2.352       2.516       2.229       2.354       2.397       2.397   
3244       7.031      17.018      17.603      15.005       8.987       8.490   
3245       5.398       6.788       7.360       6.898       7.321       8.042   
3246      27.664      29.229      29.548      27.909      26.923      21.277   
3247      15.499      11.853      17.431      14.506      12.812      10.472   

      2017-01-09  2017-01-10  2017-01-11  2017-01-12  2017-01-13  2017-01-14  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243       2.269      2.3040       2.339       2.212       2.456       2.229   
3244      10.136     10.3010      10.466       8.177      11.929       8.261   
3245       8.207      6.8355       5.464       7.545       6.432       6.001   
3246      23.452     24.6880      25.924      26.834      30.346      30.367   
3247      10.879     10.3525       9.826      11.969      14.236      15.300   

      2017-01-15  2017-01-16  2017-01-17  2017-01-18  2017-01-19  2017-01-20  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243       2.503       2.207       2.469       2.266       2.321       2.262   
3244      12.369      14.352      14.058      10.472      17.396       9.879   
3245       5.558       6.391       6.066       6.443       6.675       7.363   
3246      26.135      26.842      25.588      30.267      28.921      27.530   
3247      14.199       9.955      16.184      15.341      14.343      17.231   

      2017-01-21  2017-01-22  2017-01-23  2017-01-24  ...  2017-12-14  \
0            NaN         NaN         NaN         NaN  ...       3.509   
1            NaN         NaN         NaN         NaN  ...      22.546   
2            NaN         

# Training set 

## Converting to a supervised learning problem

* Transforming data from a time series structure into a supervised learning structure
    * restrecturing dataframe with 9 columns, 8 inputs, 1 output (meter_reading)


In [32]:
def to_supervised(df_original):

    #defining the new structure for the dataframe
    df_energy_train = pd.DataFrame(columns=["meter_id","date","energy_cluster","num_bedrooms","dwelling_type","detached","flat","semi_detached","terraced","meter_reading"])

    #iteratively populating the training set
    date = pd.date_range(datetime(2017, 1, 1), datetime(2017, 12, 31))#outside of loop as doesn't need to be recaculated

    #for each row in the dataframe
    for row_id in tqdm(range(len(df_original))):

        #getting relevant values for this row
        row = df_original.iloc[row_id]

        ### adding the various meta_data ###
        meter_id = row.meter_id
        energy_cluster = row.labels
        num_bedrooms = row.num_bedrooms
        detached = row.dwelling_type_detached_house
        flat = row.dwelling_type_flat
        semi_detached = row.dwelling_type_semi_detached_house
        terraced = row.dwelling_type_terraced_house
        dwelling_type = row.dwelling_type
        meter_reading = []
        #iterating through all the dates for this meter (skipping meter_id column up to len row - 7 metadata columns)
        for reading_index in range (1,len(row)-7):
            meter_reading.append(row[reading_index])

        #formatting this row into dataframe with timeseries patterns over rows instead of columns
        df_meter = pd.DataFrame(columns=["meter_id","date","energy_cluster","num_bedrooms","dwelling_type","detached","flat","semi_detached","terraced"])

        df_meter["date"]=date #needs to be first as it determines length of columns
        df_meter["meter_id"]=meter_id
        df_meter["energy_cluster"]=energy_cluster
        df_meter["num_bedrooms"]=num_bedrooms
        df_meter["detached"]=detached
        df_meter["flat"]=flat
        df_meter["semi_detached"]=semi_detached
        df_meter["terraced"]=terraced
        df_meter["dwelling_type"]=dwelling_type
        df_meter["meter_reading"]=meter_reading

        #finished converting this row so adding it to the dataframe
        df_energy_train = pd.concat([df_energy_train, df_meter],axis=0).reset_index(drop=True)

    return df_energy_train

##### number of previous steps we are using in our predictions
df_energy_train=to_supervised(df_energy_addInfo)
df_energy_train

## Feature engineering 
#### Dummy variables for the date
* Cyclically encoded
    * As if ordinally encoded it will imply that January (0) and December (12) are far from each other 
    


* My thinking was
    * Cyclically encode day of year as 1st Jan is close to 31st Dec
    * Cyclically encode month for the same reason
    * Not cyclically encoded day of week or month as they exist within the month and there is an ordinal relationship within that month
    * Extra dummy variable for specifying if it's a weekend or not such that it can more easily facilitate learning it


* Will experiment with different combinations of these in the RNN to see which performs the best
    * Eg does the dummy variable for saying it's a weekend or not improve performance or is it better to just use day_of_week?
        * etc.

# Inspecting number of missing meters
* 10 meters are missing so just going to use the mean for 1 day of these so their id exists in the model so it can weigh it appropriately as apposed to the meter_id not existing and it being out of scope

In [34]:
#checking the number of unique meter_id's in this
df_energy_train["meter_id"].nunique()

3248

In [35]:
# removing rows containing nans as the model won't be able to use them
df_energy_train_no_nan = df_energy_train.dropna(axis=0).reset_index(drop=True)
df_energy_train_no_nan

meter_id       date energy_cluster  \
0       0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2017-12-03              0   
1       0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2017-12-04              0   
2       0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2017-12-05              0   
3       0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2017-12-06              0   
4       0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2017-12-07              0   
...                                            ...        ...            ...   
640294  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-27              1   
640295  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-28              1   
640296  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-29              1   
640297  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-30              1   
640298  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-31              1   

        num_bedrooms   dwelling_type  detached  flat  semi_detached  terraced  \
0                2.0  terraced_house       0.0   0.0            0.0       1.0   
1                2.0  terraced_house       0.0   0.0            0.0       1.0   
2                2.0  terraced_house       0.0   0.0            0.0       1.0   
3                2.0  terraced_house       0.0   0.0            0.0       1.0   
4                2.0  terraced_house       0.0   0.0            0.0       1.0   
...              ...             ...       ...   ...            ...       ...   
640294           3.0  detached_house       1.0   0.0            0.0       0.0   
640295           3.0  detached_house       1.0   0.0            0.0       0.0   
640296           3.0  detached_house       1.0   0.0            0.0       0.0   
640297           3.0  detached_house       1.0   0.0            0.0       0.0   
640298           3.0  detached_house       1.0   0.0            0.0       0.0   

        meter_reading  
0              3.5170  
1              2.9330  
2              3.0850  
3              3.2370  
4              4.4710  
...               ...  
640294        18.7585  
640295        21.0110  
640296        19.2040  
640297        17.3970  
640298        15.2370  

[640299 rows x 10 columns]

In [36]:
#checking the number of unique meter_id's in this
df_energy_train_no_nan["meter_id"].nunique()

3238

In [37]:
#what id's are missing
missing_ids = np.setdiff1d(df_energy_train["meter_id"],df_energy_train_no_nan["meter_id"],)
missing_ids

array(['0x2ce3d582a1316db5bcfe405cbd6070268944778e',
       '0x391e6c2169c27de797ccbdf2d623365da28a6d3d',
       '0x423fa805ddb0cba9bdb4460f9a78540287eefd0e',
       '0x795fa14a925f90f11752454da39988048db6e042',
       '0x800c9648a84c44cf1a1b0a0ada6283495a69fe8b',
       '0x81fa8eddb2b09393d3719984ca5520cb50f45efd',
       '0xab6f16ec971923ef1acd4693705b28bcfe7a7115',
       '0xc73b4d676b98a29a4c3c4ed78c8f1d208b96fa11',
       '0xe2154c841409bba5b83d208c83c33938d1a02474',
       '0xe41c230484378b2312fbf1fc99d289e0b51a3c52'], dtype=object)

In [38]:
#getting a dataframe of these missing id's from the original dataset

df_missing = pd.DataFrame()

for each in missing_ids:
    df_missing = pd.concat([df_missing, df_energy_addInfo[df_energy_addInfo.meter_id == each]],axis=0).reset_index(drop=True)
    
df_missing

meter_id  2017-01-01  2017-01-02  \
0  0x2ce3d582a1316db5bcfe405cbd6070268944778e         NaN         NaN   
1  0x391e6c2169c27de797ccbdf2d623365da28a6d3d         NaN         NaN   
2  0x423fa805ddb0cba9bdb4460f9a78540287eefd0e         NaN         NaN   
3  0x795fa14a925f90f11752454da39988048db6e042         NaN         NaN   
4  0x800c9648a84c44cf1a1b0a0ada6283495a69fe8b         NaN         NaN   
5  0x81fa8eddb2b09393d3719984ca5520cb50f45efd         NaN         NaN   
6  0xab6f16ec971923ef1acd4693705b28bcfe7a7115         NaN         NaN   
7  0xc73b4d676b98a29a4c3c4ed78c8f1d208b96fa11         NaN         NaN   
8  0xe2154c841409bba5b83d208c83c33938d1a02474         NaN         NaN   
9  0xe41c230484378b2312fbf1fc99d289e0b51a3c52         NaN         NaN   

   2017-01-03  2017-01-04  2017-01-05  2017-01-06  2017-01-07  2017-01-08  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   
5         NaN         NaN         NaN         NaN         NaN         NaN   
6         NaN         NaN         NaN         NaN         NaN         NaN   
7         NaN         NaN         NaN         NaN         NaN         NaN   
8         NaN         NaN         NaN         NaN         NaN         NaN   
9         NaN         NaN         NaN         NaN         NaN         NaN   

   2017-01-09  2017-01-10  2017-01-11  2017-01-12  2017-01-13  2017-01-14  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   
5         NaN         NaN         NaN         NaN         NaN         NaN   
6         NaN         NaN         NaN         NaN         NaN         NaN   
7         NaN         NaN         NaN         NaN         NaN         NaN   
8         NaN         NaN         NaN         NaN         NaN         NaN   
9         NaN         NaN         NaN         NaN         NaN         NaN   

   2017-01-15  2017-01-16  2017-01-17  2017-01-18  2017-01-19  2017-01-20  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   
5         NaN         NaN         NaN         NaN         NaN         NaN   
6         NaN         NaN         NaN         NaN         NaN         NaN   
7         NaN         NaN         NaN         NaN         NaN         NaN   
8         NaN         NaN         NaN         NaN         NaN         NaN   
9         NaN         NaN         NaN         NaN         NaN         NaN   

   2017-01-21  2017-01-22  2017-01-23  2017-01-24  ...  2017-12-14  \
0         NaN         NaN         NaN         NaN  ...         NaN   
1         NaN         NaN         NaN         NaN  ...         NaN   
2         NaN         NaN         NaN         NaN  ...         NaN   
3         NaN         NaN         NaN         NaN  ...         NaN   
4         NaN         NaN         NaN         NaN  ...         NaN   
5         NaN         NaN         NaN         NaN  ...         NaN   
6         NaN         NaN         NaN         NaN  ...         NaN   
7         NaN         NaN         NaN         NaN  ...         NaN   
8         NaN         NaN         NaN         NaN  ...     

In [43]:
# just going to use the average for those days which only have nans
df_missing_avg=df_missing.copy()
df_missing_avg["2017-12-31"]=df_energy["2017-12-31"].mean()
df_missing_avg

meter_id  2017-01-01  2017-01-02  \
0  0x2ce3d582a1316db5bcfe405cbd6070268944778e         NaN         NaN   
1  0x391e6c2169c27de797ccbdf2d623365da28a6d3d         NaN         NaN   
2  0x423fa805ddb0cba9bdb4460f9a78540287eefd0e         NaN         NaN   
3  0x795fa14a925f90f11752454da39988048db6e042         NaN         NaN   
4  0x800c9648a84c44cf1a1b0a0ada6283495a69fe8b         NaN         NaN   
5  0x81fa8eddb2b09393d3719984ca5520cb50f45efd         NaN         NaN   
6  0xab6f16ec971923ef1acd4693705b28bcfe7a7115         NaN         NaN   
7  0xc73b4d676b98a29a4c3c4ed78c8f1d208b96fa11         NaN         NaN   
8  0xe2154c841409bba5b83d208c83c33938d1a02474         NaN         NaN   
9  0xe41c230484378b2312fbf1fc99d289e0b51a3c52         NaN         NaN   

   2017-01-03  2017-01-04  2017-01-05  2017-01-06  2017-01-07  2017-01-08  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   
5         NaN         NaN         NaN         NaN         NaN         NaN   
6         NaN         NaN         NaN         NaN         NaN         NaN   
7         NaN         NaN         NaN         NaN         NaN         NaN   
8         NaN         NaN         NaN         NaN         NaN         NaN   
9         NaN         NaN         NaN         NaN         NaN         NaN   

   2017-01-09  2017-01-10  2017-01-11  2017-01-12  2017-01-13  2017-01-14  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   
5         NaN         NaN         NaN         NaN         NaN         NaN   
6         NaN         NaN         NaN         NaN         NaN         NaN   
7         NaN         NaN         NaN         NaN         NaN         NaN   
8         NaN         NaN         NaN         NaN         NaN         NaN   
9         NaN         NaN         NaN         NaN         NaN         NaN   

   2017-01-15  2017-01-16  2017-01-17  2017-01-18  2017-01-19  2017-01-20  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   
5         NaN         NaN         NaN         NaN         NaN         NaN   
6         NaN         NaN         NaN         NaN         NaN         NaN   
7         NaN         NaN         NaN         NaN         NaN         NaN   
8         NaN         NaN         NaN         NaN         NaN         NaN   
9         NaN         NaN         NaN         NaN         NaN         NaN   

   2017-01-21  2017-01-22  2017-01-23  2017-01-24  ...  2017-12-14  \
0         NaN         NaN         NaN         NaN  ...         NaN   
1         NaN         NaN         NaN         NaN  ...         NaN   
2         NaN         NaN         NaN         NaN  ...         NaN   
3         NaN         NaN         NaN         NaN  ...         NaN   
4         NaN         NaN         NaN         NaN  ...         NaN   
5         NaN         NaN         NaN         NaN  ...         NaN   
6         NaN         NaN         NaN         NaN  ...         NaN   
7         NaN         NaN         NaN         NaN  ...         NaN   
8         NaN         NaN         NaN         NaN  ...     

In [44]:
#turning the missing values where the average has been imputed, into a supervised learning problem
df_supervised_missing = to_supervised(df_missing_avg).dropna(axis=0).reset_index(drop=True)
#concatenating this with the df_energy_train
df_energy_train_with_missing = pd.concat([df_supervised_missing, df_energy_train_no_nan],axis=0).reset_index(drop=True)

df_energy_train_with_missing

  0%|          | 0/10 [00:00<?, ?it/s]

meter_id       date energy_cluster  \
0       0x2ce3d582a1316db5bcfe405cbd6070268944778e 2017-12-31              0   
1       0x391e6c2169c27de797ccbdf2d623365da28a6d3d 2017-12-31              0   
2       0x423fa805ddb0cba9bdb4460f9a78540287eefd0e 2017-12-31              0   
3       0x795fa14a925f90f11752454da39988048db6e042 2017-12-31              0   
4       0x800c9648a84c44cf1a1b0a0ada6283495a69fe8b 2017-12-31              0   
...                                            ...        ...            ...   
640304  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-27              1   
640305  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-28              1   
640306  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-29              1   
640307  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-30              1   
640308  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-31              1   

        num_bedrooms        dwelling_type  detached  flat  semi_detached  \
0                3.0  semi_detached_house       0.0   0.0            1.0   
1                3.0  semi_detached_house       0.0   0.0            1.0   
2                3.0  semi_detached_house       0.0   0.0            1.0   
3                3.0  semi_detached_house       0.0   0.0            1.0   
4                3.0  semi_detached_house       0.0   0.0            1.0   
...              ...                  ...       ...   ...            ...   
640304           3.0       detached_house       1.0   0.0            0.0   
640305           3.0       detached_house       1.0   0.0            0.0   
640306           3.0       detached_house       1.0   0.0            0.0   
640307           3.0       detached_house       1.0   0.0            0.0   
640308           3.0       detached_house       1.0   0.0            0.0   

        terraced  meter_reading  
0            0.0      11.881495  
1            0.0      11.881495  
2            0.0      11.881495  
3            0.0      11.881495  
4            0.0      11.881495  
...          ...            ...  
640304       0.0      18.758500  
640305       0.0      21.011000  
640306       0.0      19.204000  
640307       0.0      17.397000  
640308       0.0      15.237000  

[640309 rows x 10 columns]

# Feature engineering
with dummy variables

In [45]:
#defining the new structure with the cyclically encoded date columns
df_energy_train_with_dummies = pd.DataFrame(columns=['meter_id', 'date', 'day_of_year_sin', 'day_of_year_cos', 'day_of_week', 'day_of_month', 'month_ord', 'month_sin', 'month_cos', 'is_weekend','energy_cluster', 'num_bedrooms', 'dwelling_type', 'detached',
       'flat', 'semi_detached', 'terraced',
       'meter_reading'], data=df_energy_train_with_missing)

#cyclically encoding day of year
day_of_year = df_energy_train_with_dummies["date"].dt.dayofyear.astype(int)
df_energy_train_with_dummies["day_of_year_sin"] = np.sin(2 * np.pi * day_of_year/364.0)
df_energy_train_with_dummies["day_of_year_cos"] = np.cos(2 * np.pi * day_of_year/364.0)

#cyclically encoding month
month = df_energy_train_with_dummies["date"].dt.month.astype(int)
df_energy_train_with_dummies["month_ord"] = month
df_energy_train_with_dummies["month_sin"] = np.sin(2 * np.pi * month/11.0)
df_energy_train_with_dummies["month_cos"] = np.cos(2 * np.pi * month/11.0)

#day of week not cyclically encoded as it exists within month
df_energy_train_with_dummies["day_of_week"]  = df_energy_train_with_dummies["date"].dt.dayofweek.astype(int)

#day of month not cyclically encoded as it exists within month
df_energy_train_with_dummies["day_of_month"]  = df_energy_train_with_dummies["date"].dt.day.astype(int)

#dummy variable for is_weekend
df_energy_train_with_dummies["is_weekend"]=0
df_energy_train_with_dummies.loc[df_energy_train_with_dummies.day_of_week>4, "is_weekend"] = 1

# displaying the dataframe
df_energy_train_with_dummies.tail(10)

meter_id       date  \
640299  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-22   
640300  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-23   
640301  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-24   
640302  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-25   
640303  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-26   
640304  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-27   
640305  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-28   
640306  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-29   
640307  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-30   
640308  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2017-12-31   

        day_of_year_sin  day_of_year_cos  day_of_week  day_of_month  \
640299    -1.376535e-01         0.990480            4            22   
640300    -1.205367e-01         0.992709            5            23   
640301    -1.033839e-01         0.994642            6            24   
640302    -8.620038e-02         0.996278            0            25   
640303    -6.899114e-02         0.997617            1            26   
640304    -5.176135e-02         0.998659            2            27   
640305    -3.451614e-02         0.999404            3            28   
640306    -1.726064e-02         0.999851            4            29   
640307    -2.449294e-16         1.000000            5            30   
640308     1.726064e-02         0.999851            6            31   

        month_ord  month_sin  month_cos  is_weekend energy_cluster  \
640299         12   0.540641   0.841254           0              1   
640300         12   0.540641   0.841254           1              1   
640301         12   0.540641   0.841254           1              1   
640302         12   0.540641   0.841254           0              1   
640303         12   0.540641   0.841254           0              1   
640304         12   0.540641   0.841254           0              1   
640305         12   0.540641   0.841254           0              1   
640306         12   0.540641   0.841254           0              1   
640307         12   0.540641   0.841254           1              1   
640308         12   0.540641   0.841254           1              1   

        num_bedrooms   dwelling_type  detached  flat  semi_detached  terraced  \
640299           3.0  detached_house       1.0   0.0            0.0       0.0   
640300           3.0  detached_house       1.0   0.0            0.0       0.0   
640301           3.0  detached_house       1.0   0.0            0.0       0.0   
640302           3.0  detached_house       1.0   0.0            0.0       0.0   
640303           3.0  detached_house       1.0   0.0            0.0       0.0   
640304           3.0  detached_house       1.0   0.0            0.0       0.0   
640305           3.0  detached_house       1.0   0.0            0.0       0.0   
640306           3.0  detached_house       1.0   0.0            0.0       0.0   
640307           3.0  detached_house       1.0   0.0            0.0       0.0   
640308           3.0  detached_house       1.0   0.0            0.0       0.0   

        meter_reading  
640299        14.0360  
640300        15.9900  
640301        15.5505  
640302        15.1110  
640303        16.5060  
640304        18.7585  
640305        21.0110  
640306        19.2040  
640307        17.3970  
640308        15.2370

# encoding the ID
* Encoding via binary encoding
* One hot encoding has a linear growth rate of n whereas binary encoding has a growth rate of log_2(n)
* Our n is 3248 so:
    * One hot encoding would add 3248 extra features
    * Binary encoding would add log_2(3248) extra features = 12

* Hence, allows us to uniquely describe the different id's with far fewer features and hence resulting in a much less sparse search space

## encoding for the whole dataset

In [46]:
import pandas as pd
import category_encoders as ce

binary_encoder = ce.BinaryEncoder(cols=['meter_id'])
df_energy_train_final = pd.concat([df_energy_train_with_dummies["meter_id"],binary_encoder.fit_transform(df_energy_train_with_dummies)],axis=1)

# if(os.path.exists("../Data/Prepared_Data/Supervised/preapared_consumption_all.pkl")==False):
#         df_energy_train_final.to_pickle("../Data/Prepared_Data/Supervised/consumption_all.pkl")
df_energy_train_final.to_pickle("../Data/Prepared_Data/Supervised/consumption_all.pkl")

C:\Users\psydm7\Anaconda3\envs\mle_tf\lib\site-packages\category_encoders\utils.py:21: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead
  elif pd.api.types.is_categorical(cols):


In [47]:
df_energy_train_final

meter_id  meter_id_0  meter_id_1  \
0       0x2ce3d582a1316db5bcfe405cbd6070268944778e           0           0   
1       0x391e6c2169c27de797ccbdf2d623365da28a6d3d           0           0   
2       0x423fa805ddb0cba9bdb4460f9a78540287eefd0e           0           0   
3       0x795fa14a925f90f11752454da39988048db6e042           0           0   
4       0x800c9648a84c44cf1a1b0a0ada6283495a69fe8b           0           0   
...                                            ...         ...         ...   
640304  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   
640305  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   
640306  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   
640307  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   
640308  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   

        meter_id_2  meter_id_3  meter_id_4  meter_id_5  meter_id_6  \
0                0           0           0           0           0   
1                0           0           0           0           0   
2                0           0           0           0           0   
3                0           0           0           0           0   
4                0           0           0           0           0   
...            ...         ...         ...         ...         ...   
640304           1           0           0           1           0   
640305           1           0           0           1           0   
640306           1           0           0           1           0   
640307           1           0           0           1           0   
640308           1           0           0           1           0   

        meter_id_7  meter_id_8  meter_id_9  meter_id_10  meter_id_11  \
0                0           0           0            0            0   
1                0           0           0            0            1   
2                0           0           0            0            1   
3                0           0           0            1            0   
4                0           0           0            1            0   
...            ...         ...         ...          ...          ...   
640304           1           1           0            0            0   
640305           1           1           0            0            0   
640306           1           1           0            0            0   
640307           1           1           0            0            0   
640308           1           1           0            0            0   

        meter_id_12       date  day_of_year_sin  day_of_year_cos  day_of_week  \
0                 1 2017-12-31     1.726064e-02         0.999851            6   
1                 0 2017-12-31     1.726064e-02         0.999851            6   
2                 1 2017-12-31     1.726064e-02         0.999851            6   
3                 0 2017-12-31     1.726064e-02         0.999851            6   
4                 1 2017-12-31     1.726064e-02         0.999851            6   
...             ...        ...              ...              ...          ...   
640304            0 2017-12-27    -5.176135e-02         0.998659            2   
640305            0 2017-12-28    -3.451614e-02         0.999404            3   
640306            0 2017-12-29    -1.726064e-02         0.999851            4   
640307            0 2017-12-30    -2.449294e-16         1.000000            5   
640308            0 2017-12-31     1.726064e-02         0.999851            6   

        day_of_month  month_ord  month_sin  month_cos  is_weekend  \
0                 31         12   0.540641   0.841254           1   
1                 31         12   0.540641   0.841254           1   
2                 31         12   0.540641   0.841254           1   
3                 31         12   0.540641   0.841254           1   
4                 31         12   0.540641   0.841254           1   
...         

# Prediction set 
* making an empty prediction set we will populate
* done using same strategy as above

In [48]:
### Creating a prediction set in the same structure as df_energy_addInfo so we can reuse the same code to format it ###

# making the prediction set
df_preds = pd.DataFrame(columns=["meter_id"])
df_preds["meter_id"]=df_energy_addInfo.meter_id

#adding the date columns
for i in tqdm(pd.date_range(datetime(2018,1,1),datetime(2018,12,31),freq="D"),position=0):
    df_preds[str(i.date())]=np.nan
    
#getting the missing columns
cols_to_use = df_energy_addInfo.columns.difference(df_preds.columns)[365:] #after the 365 date columns
cols_to_use = np.concatenate((cols_to_use,["meter_id"]))

df_preds = pd.merge(df_preds, df_energy_addInfo[cols_to_use], on="meter_id",how="inner")
df_preds

  0%|          | 0/365 [00:00<?, ?it/s]

meter_id  2018-01-01  2018-01-02  \
0     0xa62b9f23553ff183f61e2bf943aab3d5983d02d7         NaN         NaN   
1     0x459c834d1f6cfb5b734b82aa9f5410fa97fb70da         NaN         NaN   
2     0x4a1ed36825360a058cec2bdd409fc2459e1ce54f         NaN         NaN   
3     0x5b76d3c0e0aefc6e0a8d1d031f96388a23263407         NaN         NaN   
4     0x943ebe39ef2be6ef807c42c5a647e27112ca5b0f         NaN         NaN   
...                                          ...         ...         ...   
3243  0x7dd7a7b8ee1bec7c44b24f738c752482f6161065         NaN         NaN   
3244  0xfdaf9f857621ec06f2cf801f42a020a322835090         NaN         NaN   
3245  0xd28f2f001e0cd4d6c121a3cb2e1427207e170e18         NaN         NaN   
3246  0x47218b46abb2fcaade487a211911406dc6e13730         NaN         NaN   
3247  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd         NaN         NaN   

      2018-01-03  2018-01-04  2018-01-05  2018-01-06  2018-01-07  2018-01-08  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243         NaN         NaN         NaN         NaN         NaN         NaN   
3244         NaN         NaN         NaN         NaN         NaN         NaN   
3245         NaN         NaN         NaN         NaN         NaN         NaN   
3246         NaN         NaN         NaN         NaN         NaN         NaN   
3247         NaN         NaN         NaN         NaN         NaN         NaN   

      2018-01-09  2018-01-10  2018-01-11  2018-01-12  2018-01-13  2018-01-14  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243         NaN         NaN         NaN         NaN         NaN         NaN   
3244         NaN         NaN         NaN         NaN         NaN         NaN   
3245         NaN         NaN         NaN         NaN         NaN         NaN   
3246         NaN         NaN         NaN         NaN         NaN         NaN   
3247         NaN         NaN         NaN         NaN         NaN         NaN   

      2018-01-15  2018-01-16  2018-01-17  2018-01-18  2018-01-19  2018-01-20  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
3243         NaN         NaN         NaN         NaN         NaN         NaN   
3244         NaN         NaN         NaN         NaN         NaN         NaN   
3245         NaN         NaN         NaN         NaN         NaN         NaN   
3246         NaN         NaN         NaN         NaN         NaN         NaN   
3247         NaN         NaN         NaN         NaN         NaN         NaN   

      2018-01-21  2018-01-22  2018-01-23  2018-01-24  ...  2018-12-14  \
0            NaN         NaN         NaN         NaN  ...         NaN   
1            NaN         NaN         NaN         NaN  ...         NaN   
2            NaN         

In [49]:
def pred_to_supervised(df_original):

    #defining the new structure for the dataframe
    df_energy_pred = pd.DataFrame(columns=["meter_id","date","energy_cluster","num_bedrooms","dwelling_type","detached","flat","semi_detached","terraced","meter_reading"])

    #iteratively populating the training set
    date = pd.date_range(datetime(2018, 1, 1), datetime(2018, 12, 31))#outside of loop as doesn't need to be recaculated

    #for each row in the dataframe
    for row_id in tqdm(range(len(df_original))):

        #getting relevant values for this row
        row = df_original.iloc[row_id]

        ### adding the various meta_data ###
        meter_id = row.meter_id
        energy_cluster = row.labels
        num_bedrooms = row.num_bedrooms
        detached = row.dwelling_type_detached_house
        flat = row.dwelling_type_flat
        semi_detached = row.dwelling_type_semi_detached_house
        terraced = row.dwelling_type_terraced_house
        dwelling_type = row.dwelling_type

        #formatting this row into dataframe with timeseries patterns over rows instead of columns
        df_meter = pd.DataFrame(columns=["meter_id","date","energy_cluster","num_bedrooms","dwelling_type","detached","flat","semi_detached","terraced"])

        df_meter["date"]=date #needs to be first as it determines length of columns
        df_meter["meter_id"]=meter_id
        df_meter["energy_cluster"]=energy_cluster
        df_meter["num_bedrooms"]=num_bedrooms
        df_meter["detached"]=detached
        df_meter["flat"]=flat
        df_meter["semi_detached"]=semi_detached
        df_meter["terraced"]=terraced
        df_meter["dwelling_type"]=dwelling_type

        #finished converting this row so adding it to the dataframe
        df_energy_pred = pd.concat([df_energy_pred, df_meter],axis=0).reset_index(drop=True)

    return df_energy_pred

In [50]:
### Turning into a supervised problem ###

#number of previous steps we are using in our predictions
df_preds_super=pred_to_supervised(df_preds)
df_preds_super

  0%|          | 0/3248 [00:00<?, ?it/s]

meter_id       date energy_cluster  \
0        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-01              0   
1        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-02              0   
2        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-03              0   
3        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-04              0   
4        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-05              0   
...                                             ...        ...            ...   
1185515  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-27              1   
1185516  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-28              1   
1185517  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-29              1   
1185518  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-30              1   
1185519  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-31              1   

         num_bedrooms   dwelling_type  detached  flat  semi_detached  \
0                 2.0  terraced_house       0.0   0.0            0.0   
1                 2.0  terraced_house       0.0   0.0            0.0   
2                 2.0  terraced_house       0.0   0.0            0.0   
3                 2.0  terraced_house       0.0   0.0            0.0   
4                 2.0  terraced_house       0.0   0.0            0.0   
...               ...             ...       ...   ...            ...   
1185515           3.0  detached_house       1.0   0.0            0.0   
1185516           3.0  detached_house       1.0   0.0            0.0   
1185517           3.0  detached_house       1.0   0.0            0.0   
1185518           3.0  detached_house       1.0   0.0            0.0   
1185519           3.0  detached_house       1.0   0.0            0.0   

         terraced meter_reading  
0             1.0           NaN  
1             1.0           NaN  
2             1.0           NaN  
3             1.0           NaN  
4             1.0           NaN  
...           ...           ...  
1185515       0.0           NaN  
1185516       0.0           NaN  
1185517       0.0           NaN  
1185518       0.0           NaN  
1185519       0.0           NaN  

[1185520 rows x 10 columns]

In [51]:
### doing the date based dummies ###

#defining the new structure with the cyclically encoded date columns
df_preds_dummies = pd.DataFrame(columns=['meter_id', 'date', 'day_of_year_sin', 'day_of_year_cos', 'day_of_week', 'day_of_month', 'month_ord','month_sin', 'month_cos', 'is_weekend','energy_cluster', 'num_bedrooms', 'dwelling_type', 'detached',
       'flat', 'semi_detached', 'terraced',
       'meter_reading'], data=df_preds_super)

#cyclically encoding day of year
day_of_year = df_preds_dummies["date"].dt.dayofyear.astype(int)
df_preds_dummies["day_of_year_sin"] = np.sin(2 * np.pi * day_of_year/364.0)
df_preds_dummies["day_of_year_cos"] = np.cos(2 * np.pi * day_of_year/364.0)

#cyclically encoding month
month = df_preds_dummies["date"].dt.month.astype(int)
df_preds_dummies["month_ord"] = month
df_preds_dummies["month_sin"] = np.sin(2 * np.pi * month/11.0)
df_preds_dummies["month_cos"] = np.cos(2 * np.pi * month/11.0)

#day of week not cyclically encoded as it exists within month
df_preds_dummies["day_of_week"]  = df_preds_dummies["date"].dt.dayofweek.astype(int)

#day of month not cyclically encoded as it exists within month
df_preds_dummies["day_of_month"]  = df_preds_dummies["date"].dt.day.astype(int)

#dummy variable for is_weekend
df_preds_dummies["is_weekend"]=0
df_preds_dummies.loc[df_preds_dummies.day_of_week>4, "is_weekend"] = 1

# displaying the dataframe
df_preds_dummies

meter_id       date  \
0        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-01   
1        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-02   
2        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-03   
3        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-04   
4        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-05   
...                                             ...        ...   
1185515  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-27   
1185516  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-28   
1185517  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-29   
1185518  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-30   
1185519  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-31   

         day_of_year_sin  day_of_year_cos  day_of_week  day_of_month  \
0           1.726064e-02         0.999851            0             1   
1           3.451614e-02         0.999404            1             2   
2           5.176135e-02         0.998659            2             3   
3           6.899114e-02         0.997617            3             4   
4           8.620038e-02         0.996278            4             5   
...                  ...              ...          ...           ...   
1185515    -5.176135e-02         0.998659            3            27   
1185516    -3.451614e-02         0.999404            4            28   
1185517    -1.726064e-02         0.999851            5            29   
1185518    -2.449294e-16         1.000000            6            30   
1185519     1.726064e-02         0.999851            0            31   

         month_ord  month_sin  month_cos  is_weekend energy_cluster  \
0                1   0.540641   0.841254           0              0   
1                1   0.540641   0.841254           0              0   
2                1   0.540641   0.841254           0              0   
3                1   0.540641   0.841254           0              0   
4                1   0.540641   0.841254           0              0   
...            ...        ...        ...         ...            ...   
1185515         12   0.540641   0.841254           0              1   
1185516         12   0.540641   0.841254           0              1   
1185517         12   0.540641   0.841254           1              1   
1185518         12   0.540641   0.841254           1              1   
1185519         12   0.540641   0.841254           0              1   

         num_bedrooms   dwelling_type  detached  flat  semi_detached  \
0                 2.0  terraced_house       0.0   0.0            0.0   
1                 2.0  terraced_house       0.0   0.0            0.0   
2                 2.0  terraced_house       0.0   0.0            0.0   
3                 2.0  terraced_house       0.0   0.0            0.0   
4                 2.0  terraced_house       0.0   0.0            0.0   
...               ...             ...       ...   ...            ...   
1185515           3.0  detached_house       1.0   0.0            0.0   
1185516           3.0  detached_house       1.0   0.0            0.0   
1185517           3.0  detached_house       1.0   0.0            0.0   
1185518           3.0  detached_house       1.0   0.0            0.0   
1185519           3.0  detached_house       1.0   0.0            0.0   

         terraced meter_reading  
0             1.0           NaN  
1             1.0           NaN  
2             1.0           NaN  
3             1.0           NaN  
4             1.0           NaN  
...           ...           ...  
1185515       0.0           NaN  
1185516       0.0           NaN  
1185517       0.0           NaN  
1185518       0.0           NaN  
1185519       0.0           NaN  

[1185520 rows x 18 columns]

# encoding the ID

## encoding for the whole dataset

In [52]:
df_preds_dummies

meter_id       date  \
0        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-01   
1        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-02   
2        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-03   
3        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-04   
4        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7 2018-01-05   
...                                             ...        ...   
1185515  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-27   
1185516  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-28   
1185517  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-29   
1185518  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-30   
1185519  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd 2018-12-31   

         day_of_year_sin  day_of_year_cos  day_of_week  day_of_month  \
0           1.726064e-02         0.999851            0             1   
1           3.451614e-02         0.999404            1             2   
2           5.176135e-02         0.998659            2             3   
3           6.899114e-02         0.997617            3             4   
4           8.620038e-02         0.996278            4             5   
...                  ...              ...          ...           ...   
1185515    -5.176135e-02         0.998659            3            27   
1185516    -3.451614e-02         0.999404            4            28   
1185517    -1.726064e-02         0.999851            5            29   
1185518    -2.449294e-16         1.000000            6            30   
1185519     1.726064e-02         0.999851            0            31   

         month_ord  month_sin  month_cos  is_weekend energy_cluster  \
0                1   0.540641   0.841254           0              0   
1                1   0.540641   0.841254           0              0   
2                1   0.540641   0.841254           0              0   
3                1   0.540641   0.841254           0              0   
4                1   0.540641   0.841254           0              0   
...            ...        ...        ...         ...            ...   
1185515         12   0.540641   0.841254           0              1   
1185516         12   0.540641   0.841254           0              1   
1185517         12   0.540641   0.841254           1              1   
1185518         12   0.540641   0.841254           1              1   
1185519         12   0.540641   0.841254           0              1   

         num_bedrooms   dwelling_type  detached  flat  semi_detached  \
0                 2.0  terraced_house       0.0   0.0            0.0   
1                 2.0  terraced_house       0.0   0.0            0.0   
2                 2.0  terraced_house       0.0   0.0            0.0   
3                 2.0  terraced_house       0.0   0.0            0.0   
4                 2.0  terraced_house       0.0   0.0            0.0   
...               ...             ...       ...   ...            ...   
1185515           3.0  detached_house       1.0   0.0            0.0   
1185516           3.0  detached_house       1.0   0.0            0.0   
1185517           3.0  detached_house       1.0   0.0            0.0   
1185518           3.0  detached_house       1.0   0.0            0.0   
1185519           3.0  detached_house       1.0   0.0            0.0   

         terraced meter_reading  
0             1.0           NaN  
1             1.0           NaN  
2             1.0           NaN  
3             1.0           NaN  
4             1.0           NaN  
...           ...           ...  
1185515       0.0           NaN  
1185516       0.0           NaN  
1185517       0.0           NaN  
1185518       0.0           NaN  
1185519       0.0           NaN  

[1185520 rows x 18 columns]

In [53]:
# joining based on meter_id to getting the missing binary_encoded meter_id's
cols_to_use = df_energy_train_final.columns.difference(df_preds_dummies.columns)
cols_to_use = np.concatenate((cols_to_use,["meter_id"]))
df_preds_temp = pd.merge(df_preds_dummies, df_energy_train_final.drop_duplicates("meter_id")[cols_to_use], on="meter_id",how="inner")

#restructuring columns to be in expected order
df_preds_final=pd.DataFrame(columns=df_energy_train_final.columns)
for column_name in df_preds_final.columns:
    df_preds_final[column_name]=df_preds_temp[column_name]

#saving this to pickle
df_preds_final.to_pickle("../Data/Prepared_Data/Supervised/consumption_all_preds.pkl")

df_preds_final

meter_id  meter_id_0  meter_id_1  \
0        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7           0           0   
1        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7           0           0   
2        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7           0           0   
3        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7           0           0   
4        0xa62b9f23553ff183f61e2bf943aab3d5983d02d7           0           0   
...                                             ...         ...         ...   
1185515  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   
1185516  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   
1185517  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   
1185518  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   
1185519  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd           0           1   

         meter_id_2  meter_id_3  meter_id_4  meter_id_5  meter_id_6  \
0                 0           0           0           0           0   
1                 0           0           0           0           0   
2                 0           0           0           0           0   
3                 0           0           0           0           0   
4                 0           0           0           0           0   
...             ...         ...         ...         ...         ...   
1185515           1           0           0           1           0   
1185516           1           0           0           1           0   
1185517           1           0           0           1           0   
1185518           1           0           0           1           0   
1185519           1           0           0           1           0   

         meter_id_7  meter_id_8  meter_id_9  meter_id_10  meter_id_11  \
0                 0           0           1            0            1   
1                 0           0           1            0            1   
2                 0           0           1            0            1   
3                 0           0           1            0            1   
4                 0           0           1            0            1   
...             ...         ...         ...          ...          ...   
1185515           1           1           0            0            0   
1185516           1           1           0            0            0   
1185517           1           1           0            0            0   
1185518           1           1           0            0            0   
1185519           1           1           0            0            0   

         meter_id_12       date  day_of_year_sin  day_of_year_cos  \
0                  1 2018-01-01     1.726064e-02         0.999851   
1                  1 2018-01-02     3.451614e-02         0.999404   
2                  1 2018-01-03     5.176135e-02         0.998659   
3                  1 2018-01-04     6.899114e-02         0.997617   
4                  1 2018-01-05     8.620038e-02         0.996278   
...              ...        ...              ...              ...   
1185515            0 2018-12-27    -5.176135e-02         0.998659   
1185516            0 2018-12-28    -3.451614e-02         0.999404   
1185517            0 2018-12-29    -1.726064e-02         0.999851   
1185518            0 2018-12-30    -2.449294e-16         1.000000   
1185519            0 2018-12-31     1.726064e-02         0.999851   

         day_of_week  day_of_month  month_ord  month_sin  month_cos  \
0                  0             1          1   0.540641   0.841254   
1                  1             2          1   0.540641   0.841254   
2                  2             3          1   0.540641   0.841254   
3                  3             4          1   0.540641   0.841254   
4                  4             5          1   0.540641   0.841254   
...              ...           ...        ...        ...        ...   
1185515            3            27    

#### Note:

* I did consider using other dummies such as "meter_mean" and information on the max/min and quartiles
* However I opted not to do this as it would likely lead to overfitting of the training data
    * Eg if one meter only has Decembers readings; it's meter-mean will only be decembers mean
    * Hence it may use that well to predict the dates in December but then when it is used to forecast all the other months it will perform very poorly
    
    